arch: arm: Differentiate between stack align and guard size

agross-oss opened the pull request:
This patch fixes a bug related to the stack guards.  If guards are
disabled, don't add anything to the size.  The alignment is still
required as the minimum alignment without guard is 4 bytes, and with
guard is 32 bytes.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

galak commented on include/arch/arm/arch.h:
```diff
@@ -43,9 +43,11 @@ extern "C" {
 #if defined(CONFIG_MPU_STACK_GUARD)
-#define STACK_ALIGN  32
+#define STACK_ALIGN		32
+#define STACK_GUARD_SIZE	4
```
Can we add a comment about what STACK_GUARD_SIZE is.

I assume this is carving out 4 bytes for the "poison" mpu region

agross-linaro replied:
sure thing.  i'll repush in a few

chunlinhan replied:
Hi,

Why do we set STACK_GUARD_SIZE to 4 when stack guard is enabled?
Set it to 32 might be more reasonable since the minimum size of MPU region should be 32.
And set STACK_GUARD_SIZE to 4 will also cause problem since in
configure_mpu_stack_guard(), we will subtract STACK_ALIGN from stack start address.

agross-linaro replied:
The stack we are defining is some user defined length.  The minimum alignment for any MPU region is 32 bytes, regardless of the unique capabilities of the MPU.  The minimum region size is 32 for ARM MPUs but not for one of the NXP ones

So you're right we have to account for this.

galak commented:
Just curious, how does arch/arm/include/cortex_m/stack.h play with all this and the definition of STACK_ALIGN_SIZE?

agross-linaro commented:
The STACK_ALIGN should be set to the STACK_ALIGN_SIZE for non mpu.  I need to add a comment to say that the min align size is 32 for mpu enabled stacks due to mpu constraints, in addition to the guard size comment

tyrantlizard77 reviewed (commented):
I think it might be better to distinguish the required alignment of the *base* of the stack, and the initial stack pointer, seems like STACK_ALIGN is doing double duty here.

For example, on x86, the base of a stack has to be aligned to a 4K page, but an initial stack pointer just has to be 4-byte aligned. So we have separate STACK_ALIGN and STACK_BASE_ALIGN macros for these two cases.

galak reviewed (changes requested):
Can you add the comment and make some of the clean ups on STACK_ALIGN & STACK_ALIGN_SIZE & MPU align.

Commit:
arch: arm: Remove redundant check and fix rounding

This patch removes the redundant stack alignment check being done.  The
stack definition macros enforce the alignment requirements via the
__align() directives.

In addition, fix the rounding down of the psp to be correct.  The
actual initial stack pointer is the end of the stack minus the size of
the __esf structure.  Rounding down after the subtraction will get us
to the correct offset.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
arm: mpu: Account for stack guard size correctly

This patch fixes a couple of issues with the stack guard size and
properly constructs the STACK_ALIGN and STACK_ALIGN_SIZE definitions.

The ARM AAPCS requires that the stack pointers be 8 byte aligned.  The
STACK_ALIGN_SIZE definition is meant to contain the stack pointer
alignment requirements.  This is the required alignment at public API
boundaries (ie stack frames).

The STACK_ALIGN definition is the required alignment for the start
address for stack buffer storage.  STACK_ALIGN is used to validate
the allocation sizes for stack buffers.

The MPU_GUARD_ALIGN_AND_SIZE definition is the minimum alignment and
size for the MPU.  The minimum size and alignment just so happen to be
32 bytes for vanilla ARM MPU implementations.

When defining stack buffers, the stack guard alignment requirements
must be taken into consideration when allocating the stack memory.
The __align() must be filled in with either STACK_ALIGN_SIZE or the
align/size of the MPU stack guard.  The align/size for the guard region
will be 0 when CONFIG_MPU_STACK_GUARD is not set, and 32 bytes when it
is.

The _ARCH_THREAD_STACK_XXXXXX APIs need to know the minimum alignment
requirements for the stack buffer memory and the stack guard size to
correctly allocate and reference the stack memory.  This is reflected
in the macros with the use of the STACK_ALIGN definition and the
MPU_GUARD_ALIGN_AND_SIZE definition.

Signed-off-by: Andy Gross <andy.gross@linaro.org>
